File: backend/app/api/patient_charts.py
from __future__ import annotations

from datetime import datetime, date, timedelta
from types import SimpleNamespace
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import func, text
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.models.patient import Patient
from app.models.certification import Certification
from app.models.f2f_encounter import F2FEncounter
from app.models.task import Task
from app.models.user import User
from app.models.communications_log import CommunicationsLog
from app.models.incident_report import IncidentReport
from app.models.clinical_note import ClinicalNote
from app.models.patient_assignment import PatientAssignment
from app.models.enums import Discipline, TaskStatus, TaskType
from app.services.bereavement_aggregation_engine import (
    BereavementAggregationEngine,
    BereavementNoteInput,
)
from app.services.eligibility.engine import evaluate_hospice_eligibility


router = APIRouter(prefix="/patient-charts", tags=["patient-charts"])

bereavement_engine = BereavementAggregationEngine()


def _load_patient(db: Session, patient_id: UUID) -> Patient:
    patient = db.query(Patient).filter(Patient.id == patient_id).first()
    if not patient:
        raise HTTPException(status_code=404, detail="Patient not found")
    return patient


def _patient_namespace(patient: Patient) -> SimpleNamespace:
    return SimpleNamespace(
        id=patient.id,
        tenant_id=patient.tenant_id,
        primary_diagnosis_description=patient.primary_diagnosis,
        primary_diagnosis_code=None,
        secondary_diagnoses=[],
    )


def _serialize_datetime(value):
    return value.isoformat() if value else None


def _serialize_date(value):
    return value.isoformat() if value else None


def _enum_value(value):
    return value.value if hasattr(value, "value") else value


def _base_patient(patient: Patient) -> dict:
    return {
        "id": str(patient.id),
        "mrn": patient.mrn,
        "full_name": patient.full_name,
        "primary_diagnosis": patient.primary_diagnosis,
        "status": patient.status,
        "acuity_state": patient.acuity_state,
        "admission_status": patient.admission_status,
        "hospice_election_date": _serialize_date(patient.hospice_election_date),
        "soc_date": _serialize_datetime(patient.soc_date),
    }


def _count_by(items: list[dict], key: str) -> dict[str, int]:
    counts: dict[str, int] = {}
    for item in items:
        value = item.get(key) or "Unknown"
        counts[value] = counts.get(value, 0) + 1
    return counts


def _load_visit_rows(db: Session, patient_id: UUID, *, limit: int | None = None, ascending: bool = True):
    order = "ASC" if ascending else "DESC"
    sql = f"""
        SELECT
            v.id,
            v.visit_datetime,
            v.visit_type,
            v.visit_discipline,
            v.status,
            v.is_supervisory,
            u.full_name AS provider_name
        FROM visits v
        JOIN users u ON u.id = v.provider_id
        WHERE v.patient_id = :patient_id
        ORDER BY v.visit_datetime {order}
    """
    if limit is not None:
        sql += " LIMIT :limit"

    params = {"patient_id": patient_id}
    if limit is not None:
        params["limit"] = limit

    return db.execute(text(sql), params).mappings().all()


@router.get("/{patient_id}/summary")
def get_patient_chart_summary(patient_id: UUID, db: Session = Depends(get_db)):
    patient = _load_patient(db, patient_id)

    communication_rows = (
        db.query(CommunicationsLog)
        .filter(CommunicationsLog.patient_id == patient.id)
        .order_by(CommunicationsLog.event_time.desc(), CommunicationsLog.created_at.desc())
        .limit(5)
        .all()
    )

    incident_rows = (
        db.query(IncidentReport)
        .filter(IncidentReport.patient_id == patient.id)
        .order_by(IncidentReport.incident_date.desc(), IncidentReport.created_at.desc())
        .limit(5)
        .all()
    )

    visit_rows = _load_visit_rows(db, patient.id, limit=8, ascending=False)

    assignment_rows = (
        db.query(PatientAssignment, User.full_name.label("staff_name"))
        .join(User, User.id == PatientAssignment.user_id)
        .filter(PatientAssignment.patient_id == patient.id)
        .filter(PatientAssignment.active.is_(True))
        .order_by(PatientAssignment.is_primary.desc(), PatientAssignment.assigned_at.desc())
        .all()
    )

    return {
        "patient": _base_patient(patient),
        "care_team": [
            {
                "discipline": row.PatientAssignment.discipline.value if hasattr(row.PatientAssignment.discipline, "value") else str(row.PatientAssignment.discipline),
                "staff_name": row.staff_name,
                "primary": bool(row.PatientAssignment.is_primary),
                "status": row.PatientAssignment.status,
                "service_area": row.PatientAssignment.service_area,
            }
            for row in assignment_rows
        ],
        "recent_visits": [
            {
                "id": str(visit["id"]),
                "visit_datetime": _serialize_datetime(visit["visit_datetime"]),
                "visit_type": visit["visit_type"],
                "discipline": visit["visit_discipline"],
                "status": visit["status"],
                "provider_name": visit["provider_name"],
            }
            for visit in visit_rows
        ],
        "communication_summary": {
            "total": db.query(func.count(CommunicationsLog.id))
            .filter(CommunicationsLog.patient_id == patient.id)
            .scalar()
            or 0,
            "latest": [
                {
                    "id": str(row.id),
                    "event_type": row.event_type,
                    "focus_area": row.focus_area,
                    "event_time": _serialize_datetime(row.event_time),
                    "summary": row.summary,
                    "status": row.status,
                }
                for row in communication_rows
            ],
        },
        "incident_summary": {
            "total": db.query(func.count(IncidentReport.id))
            .filter(IncidentReport.patient_id == patient.id)
            .scalar()
            or 0,
            "latest": [
                {
                    "id": str(row.id),
                    "incident_type": row.incident_type,
                    "incident_severity": row.incident_severity,
                    "incident_date": _serialize_date(row.incident_date),
                    "reported_date": _serialize_date(row.reported_date),
                    "narrative": row.narrative,
                }
                for row in incident_rows
            ],
        },
        "compliance_summary": get_patient_compliance(patient_id, db),
        "volunteer_summary": get_patient_volunteer_schedule(patient_id, db),
    }


@router.get("/{patient_id}/physician")
def get_patient_physician_summary(patient_id: UUID, db: Session = Depends(get_db)):
    patient = _load_patient(db, patient_id)

    certifications = (
        db.query(Certification)
        .filter(Certification.patient_id == patient.id)
        .order_by(Certification.signed_at.desc())
        .all()
    )

    f2f_encounters = (
        db.query(F2FEncounter)
        .filter(F2FEncounter.patient_id == patient.id)
        .order_by(F2FEncounter.encounter_date.desc(), F2FEncounter.created_at.desc())
        .all()
    )

    return {
        "patient": _base_patient(patient),
        "metrics": [
            {"label": "CTI / Certifications", "value": len(certifications)},
            {"label": "F2F encounters", "value": len(f2f_encounters)},
            {
                "label": "Finalized F2F",
                "value": sum(1 for row in f2f_encounters if (row.status or "").upper() == "FINALIZED"),
            },
        ],
        "cti": [
            {
                "id": str(row.id),
                "cert_type": row.cert_type,
                "signed_at": _serialize_datetime(row.signed_at),
                "effective_date": _serialize_date(row.effective_date),
                "signed_by_role": row.signed_by_role,
                "status": row.status,
            }
            for row in certifications
        ],
        "f2f": [
            {
                "id": str(row.id),
                "encounter_date": _serialize_date(row.encounter_date),
                "performed_by_role": row.performed_by_role,
                "status": row.status,
                "finalized_at": _serialize_datetime(row.finalized_at),
                "summary": row.summary,
                "clinical_decline_summary": row.clinical_decline_summary,
            }
            for row in f2f_encounters
        ],
    }


@router.get("/{patient_id}/communication-log")
def get_patient_communication_log(patient_id: UUID, db: Session = Depends(get_db)):
    _load_patient(db, patient_id)

    rows = (
        db.query(CommunicationsLog)
        .filter(CommunicationsLog.patient_id == patient_id)
        .order_by(CommunicationsLog.event_time.desc(), CommunicationsLog.created_at.desc())
        .all()
    )

    return {
        "total": len(rows),
        "counts_by_type": _count_by(
            [{"event_type": row.event_type} for row in rows],
            "event_type",
        ),
        "entries": [
            {
                "id": str(row.id),
                "event_type": row.event_type,
                "focus_area": row.focus_area,
                "event_time": _serialize_datetime(row.event_time),
                "summary": row.summary,
                "details": row.details,
                "status": row.status,
                "created_at": _serialize_datetime(row.created_at),
            }
            for row in rows
        ],
    }


@router.get("/{patient_id}/incident-occurrence")
def get_patient_incident_occurrences(patient_id: UUID, db: Session = Depends(get_db)):
    _load_patient(db, patient_id)

    rows = (
        db.query(IncidentReport)
        .filter(IncidentReport.patient_id == patient_id)
        .order_by(IncidentReport.incident_date.desc(), IncidentReport.created_at.desc())
        .all()
    )

    items = [
        {
            "id": str(row.id),
            "incident_type": row.incident_type,
            "incident_severity": row.incident_severity,
            "incident_date": _serialize_date(row.incident_date),
            "reported_date": _serialize_date(row.reported_date),
            "incident_time": row.incident_time.isoformat() if row.incident_time else None,
            "reported_by": row.reported_by,
            "witnessed_by": row.witnessed_by,
            "place": row.place,
            "area": row.area,
            "surface": row.surface,
            "medication_used": row.medication_used,
            "activity_at_time": row.activity_at_time,
            "injury_level": row.injury_level,
            "injury_type": row.injury_type,
            "other_injury_text": row.other_injury_text,
            "narrative": row.narrative,
            "signed_at": _serialize_datetime(row.signed_at),
        }
        for row in rows
    ]

    return {
        "total": len(items),
        "counts_by_type": _count_by(items, "incident_type"),
        "counts_by_severity": _count_by(items, "incident_severity"),
        "items": items,
    }


@router.get("/{patient_id}/bereavement")
def get_patient_bereavement(patient_id: UUID, db: Session = Depends(get_db)):
    patient = _load_patient(db, patient_id)

    note_rows = (
        db.query(ClinicalNote)
        .filter(ClinicalNote.patient_id == patient.id)
        .order_by(ClinicalNote.created_at.desc())
        .all()
    )

    comm_rows = (
        db.query(CommunicationsLog)
        .filter(CommunicationsLog.patient_id == patient.id)
        .order_by(CommunicationsLog.event_time.desc(), CommunicationsLog.created_at.desc())
        .all()
    )

    notes = [
        BereavementNoteInput(
            patient_id=patient.id,
            note_id=row.id,
            discipline=(row.discipline or ""),
            text=(row.content or ""),
        )
        for row in note_rows
    ]
    result = bereavement_engine.detect(notes)

    relevant_comms = [
        {
            "id": str(row.id),
            "event_type": row.event_type,
            "event_time": _serialize_datetime(row.event_time),
            "summary": row.summary,
            "status": row.status,
        }
        for row in comm_rows
        if "bereav" in (row.event_type or "").lower()
        or "grief" in (row.summary or "").lower()
        or "loss" in (row.summary or "").lower()
        or "mourning" in (row.summary or "").lower()
    ]

    return {
        "patient": _base_patient(patient),
        "aggregation": {
            "rn_present": result.rn_present,
            "sw_present": result.sw_present,
            "chaplain_present": result.chaplain_present,
            "reason_codes": result.reason_codes,
            "source_notes": [str(note_id) for note_id in result.source_notes],
        },
        "supporting_notes": [
            {
                "id": str(row.id),
                "discipline": row.discipline,
                "form_key": row.form_key,
                "created_at": _serialize_datetime(row.created_at),
                "content": row.content,
            }
            for row in note_rows[:10]
            if row.content
        ],
        "supporting_communications": relevant_comms[:10],
    }


@router.get("/{patient_id}/compliance")
def get_patient_compliance(patient_id: UUID, db: Session = Depends(get_db)):
    patient = _load_patient(db, patient_id)
    patient_ctx = _patient_namespace(patient)

    eligibility = evaluate_hospice_eligibility(patient_ctx, date.today().isoformat())

    task_rows = (
        db.query(Task)
        .filter(Task.patient_id == patient.id)
        .all()
    )

    note_rows = (
        db.query(ClinicalNote)
        .filter(ClinicalNote.patient_id == patient.id)
        .order_by(ClinicalNote.created_at.desc())
        .all()
    )

    hope_notes = [row for row in note_rows if (row.form_key or "").upper().startswith("HOPE")]
    poc_notes = [row for row in note_rows if (row.form_key or "").upper().find("POC") >= 0]
    f2f_notes = [row for row in note_rows if (row.form_key or "").upper().find("F2F") >= 0]

    task_counts = {
        "pending": sum(1 for row in task_rows if _enum_value(row.status) == "PENDING"),
        "overdue": sum(1 for row in task_rows if _enum_value(row.status) == "OVERDUE"),
        "completed": sum(1 for row in task_rows if _enum_value(row.status) == "COMPLETED"),
    }

    open_issues = []
    if task_counts["pending"] or task_counts["overdue"]:
        open_issues.append("Open workflow tasks")
    if not hope_notes:
        open_issues.append("HOPE notes not yet documented")
    if not f2f_notes:
        open_issues.append("F2F note not yet documented")

    return {
        "patient": _base_patient(patient),
        "eligibility": eligibility,
        "task_counts": task_counts,
        "note_counts": {
            "total": len(note_rows),
            "hope": len(hope_notes),
            "poc": len(poc_notes),
            "f2f": len(f2f_notes),
        },
        "hope_status": "READY" if hope_notes else "NOT_STARTED",
        "qies_status": "READY" if eligibility.get("eligible") and task_counts["pending"] == 0 else "NEEDS_REVIEW",
        "open_issues": open_issues,
        "recent_notes": [
            {
                "id": str(row.id),
                "form_key": row.form_key,
                "note_type": row.note_type,
                "status": row.status,
                "created_at": _serialize_datetime(row.created_at),
                "content": row.content,
            }
            for row in note_rows[:5]
        ],
    }


@router.get("/{patient_id}/volunteer-scheduling")
def get_patient_volunteer_schedule(patient_id: UUID, db: Session = Depends(get_db)):
    patient = _load_patient(db, patient_id)

    visit_rows = _load_visit_rows(db, patient.id, ascending=True)

    assignment_rows = (
        db.query(PatientAssignment, User.full_name.label("staff_name"))
        .join(User, User.id == PatientAssignment.user_id)
        .filter(PatientAssignment.patient_id == patient.id)
        .filter(
            PatientAssignment.discipline.in_([
                Discipline.VOLUNTEER,
                Discipline.VOLUNTEER_COORDINATOR,
            ])
        )
        .order_by(PatientAssignment.assigned_at.desc())
        .all()
    )

    task_rows = (
        db.query(Task)
        .filter(Task.patient_id == patient.id)
        .filter(Task.task_type.in_([TaskType.HUV, TaskType.OTHER]))
        .order_by(Task.due_date.asc())
        .all()
    )

    return {
        "patient": _base_patient(patient),
        "visits": [
            {
                "id": str(visit["id"]),
                "visit_datetime": _serialize_datetime(visit["visit_datetime"]),
                "visit_type": visit["visit_type"],
                "visit_discipline": visit["visit_discipline"],
                "status": visit["status"],
                "provider_name": visit["provider_name"],
                "is_supervisory": bool(visit["is_supervisory"]),
            }
            for visit in visit_rows
        ],
        "assignments": [
            {
                "id": str(row.PatientAssignment.id),
                "discipline": row.PatientAssignment.discipline.value if hasattr(row.PatientAssignment.discipline, "value") else str(row.PatientAssignment.discipline),
                "staff_name": row.staff_name,
                "primary": bool(row.PatientAssignment.is_primary),
                "service_area": row.PatientAssignment.service_area,
                "status": row.PatientAssignment.status,
                "assigned_at": _serialize_datetime(row.PatientAssignment.assigned_at),
            }
            for row in assignment_rows
        ],
        "task_slots": [
            {
                "id": str(row.id),
                "task_type": _enum_value(row.task_type),
                "status": _enum_value(row.status),
                "due_date": _serialize_date(row.due_date),
                "assigned_user_id": str(row.assigned_user_id) if row.assigned_user_id else None,
                "assigned_role": row.assigned_role,
                "alert_reason": row.alert_reason,
            }
            for row in task_rows
        ],
    }

File: sns-emr-frontend/src/api/patientCharts.ts
import api from "./client";

export type PatientSummaryResponse = {
  patient: {
    id: string;
    mrn: string;
    full_name: string;
    primary_diagnosis: string;
    status: string;
    acuity_state: string;
    admission_status: string;
    hospice_election_date: string | null;
    soc_date: string | null;
  };
  care_team: Array<{
    discipline: string;
    staff_name: string;
    primary: boolean;
    status: string;
    service_area: string | null;
  }>;
  recent_visits: Array<{
    id: string;
    visit_datetime: string | null;
    visit_type: string;
    discipline: string | null;
    status: string;
    provider_name: string;
  }>;
  communication_summary: {
    total: number;
    latest: Array<{
      id: string;
      event_type: string;
      focus_area: string | null;
      event_time: string | null;
      summary: string;
      status: string;
    }>;
  };
  incident_summary: {
    total: number;
    latest: Array<{
      id: string;
      incident_type: string;
      incident_severity: string;
      incident_date: string | null;
      reported_date: string | null;
      narrative: string | null;
    }>;
  };
  compliance_summary: ComplianceResponse;
  volunteer_summary: VolunteerSchedulingResponse;
};

export type PhysicianSummaryResponse = {
  patient: PatientSummaryResponse["patient"];
  metrics: Array<{
    label: string;
    value: number;
  }>;
  cti: Array<{
    id: string;
    cert_type: string;
    signed_at: string | null;
    effective_date: string | null;
    signed_by_role: string;
    status: string;
  }>;
  f2f: Array<{
    id: string;
    encounter_date: string | null;
    performed_by_role: string;
    status: string;
    finalized_at: string | null;
    summary: string | null;
    clinical_decline_summary: string | null;
  }>;
};

export type CommunicationLogResponse = {
  total: number;
  counts_by_type: Record<string, number>;
  entries: Array<{
    id: string;
    event_type: string;
    focus_area: string | null;
    event_time: string | null;
    summary: string;
    details: Record<string, unknown> | null;
    status: string;
    created_at: string | null;
  }>;
};

export type IncidentOccurrenceResponse = {
  total: number;
  counts_by_type: Record<string, number>;
  counts_by_severity: Record<string, number>;
  items: Array<{
    id: string;
    incident_type: string;
    incident_severity: string;
    incident_date: string | null;
    reported_date: string | null;
    incident_time: string | null;
    reported_by: string | null;
    witnessed_by: string | null;
    place: string | null;
    area: string | null;
    surface: string | null;
    medication_used: string | null;
    activity_at_time: string | null;
    injury_level: string | null;
    injury_type: string | null;
    other_injury_text: string | null;
    narrative: string | null;
    signed_at: string | null;
  }>;
};

export type BereavementResponse = {
  patient: PatientSummaryResponse["patient"];
  aggregation: {
    rn_present: boolean;
    sw_present: boolean;
    chaplain_present: boolean;
    reason_codes: string[];
    source_notes: string[];
  };
  supporting_notes: Array<{
    id: string;
    discipline: string | null;
    form_key: string | null;
    created_at: string | null;
    content: string;
  }>;
  supporting_communications: Array<{
    id: string;
    event_type: string;
    event_time: string | null;
    summary: string;
    status: string;
  }>;
};

export type ComplianceResponse = {
  patient: PatientSummaryResponse["patient"];
  eligibility: Record<string, unknown>;
  task_counts: {
    pending: number;
    overdue: number;
    completed: number;
  };
  note_counts: {
    total: number;
    hope: number;
    poc: number;
    f2f: number;
  };
  hope_status: string;
  qies_status: string;
  open_issues: string[];
  recent_notes: Array<{
    id: string;
    form_key: string | null;
    note_type: string;
    status: string | null;
    created_at: string | null;
    content: string;
  }>;
};

export type VolunteerSchedulingResponse = {
  patient: PatientSummaryResponse["patient"];
  visits: Array<{
    id: string;
    visit_datetime: string | null;
    visit_type: string;
    visit_discipline: string | null;
    status: string;
    provider_name: string;
    is_supervisory: boolean;
  }>;
  assignments: Array<{
    id: string;
    discipline: string;
    staff_name: string;
    primary: boolean;
    service_area: string | null;
    status: string;
    assigned_at: string | null;
  }>;
  task_slots: Array<{
    id: string;
    task_type: string;
    status: string;
    due_date: string | null;
    assigned_user_id: string | null;
    assigned_role: string | null;
    alert_reason: string | null;
  }>;
};

export type AssessmentHistoryResponse = {
  patient_id: string;
  items: Array<{
    record_id: string;
    source_table: string;
    discipline: string;
    assessment_type: string;
    phase_hint: string | null;
    visit_date: string | null;
    status: string;
    locked: boolean;
    locked_at: string | null;
    locked_by: string | null;
    created_at: string | null;
    updated_at: string | null;
    record_url_hint: {
      section: string;
      assessment_id: string;
      source_table?: string;
    };
  }>;
  total: number;
  limit: number;
  offset: number;
  sort_order: "asc" | "desc";
  filters: {
    discipline: string | null;
    assessment_type: string | null;
    status: string | null;
    from_date: string | null;
    to_date: string | null;
  };
};


export async function fetchPatientSummary(patientId: string) {
  const response = await api.get<PatientSummaryResponse>(`/patient-charts/${patientId}/summary`);
  return response.data;
}

export async function fetchPhysicianSummary(patientId: string) {
  const response = await api.get<PhysicianSummaryResponse>(`/patient-charts/${patientId}/physician`);
  return response.data;
}

export async function fetchCommunicationLog(patientId: string) {
  const response = await api.get<CommunicationLogResponse>(`/patient-charts/${patientId}/communication-log`);
  return response.data;
}

export async function fetchIncidentOccurrence(patientId: string) {
  const response = await api.get<IncidentOccurrenceResponse>(`/patient-charts/${patientId}/incident-occurrence`);
  return response.data;
}

export async function fetchBereavement(patientId: string) {
  const response = await api.get<BereavementResponse>(`/patient-charts/${patientId}/bereavement`);
  return response.data;
}

export async function fetchCompliance(patientId: string) {
  const response = await api.get<ComplianceResponse>(`/patient-charts/${patientId}/compliance`);
  return response.data;
}

export async function fetchVolunteerScheduling(patientId: string) {
  const response = await api.get<VolunteerSchedulingResponse>(`/patient-charts/${patientId}/volunteer-scheduling`);
  return response.data;
}

export async function fetchAssessmentHistory(
  patientId: string,
  params?: Partial<AssessmentHistoryResponse["filters"]> & { limit?: number; offset?: number; sort_order?: "asc" | "desc" }
) {
  const response = await api.get<AssessmentHistoryResponse>(`/patients/${patientId}/assessment-history`, { params });
  return response.data;
}

export type DischargeChecklist = {
  plan_reviewed: boolean | null;
  notified: boolean | null;
  explained: boolean | null;
  readmission_explained: boolean | null;
  medication_instruction: boolean | null;
  contact_provided: boolean | null;
  referral_provided: boolean | null;
};

export type DischargeState = {
  patient_id: string;
  patient_status: string;
  admission_id: string | null;
  admission_status: string | null;
  discharged: boolean;
  discharge_date: string | null;
  discharge_reason: string | null;
  discharge_initiated_by: string | null;
  discharge_projected_date: string | null;
  discharge_comments: string | null;
  checklist: DischargeChecklist;
  reason_codes: Record<string, string>;
};

export async function fetchDischargePlanning(patientId: string) {
  const response = await api.get<DischargeState>(`/patients/${patientId}/discharge`);
  return response.data;
}

export async function updateDischargePlanning(
  patientId: string,
  payload: Partial<{
    discharge_projected_date: string | null;
    discharge_comments: string | null;
    discharge_plan_reviewed: boolean;
    discharge_notified: boolean;
    discharge_explained: boolean;
    discharge_readmission_explained: boolean;
    discharge_medication_instruction: boolean;
    discharge_contact_provided: boolean;
    discharge_referral_provided: boolean;
  }>
) {
  const response = await api.put<DischargeState>(`/patients/${patientId}/discharge`, payload);
  return response.data;
}

export async function finalizePatientDischarge(
  patientId: string,
  payload: { discharge_date: string; reason_code: string; notes?: string }
) {
  const response = await api.post(`/patients/${patientId}/discharge/finalize`, payload);
  return response.data;
}
